File: android/src/main/java/rabianr/capacitor/linelogin/LineLoginPlugin.java
package rabianr.capacitor.linelogin;

import androidx.activity.result.ActivityResult;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "LineLogin")
public class LineLoginPlugin extends Plugin {

    private LineLogin implementation;

    @Override
    public void load() {
        implementation = new LineLogin(getActivity().getApplicationContext());
    }

    @PluginMethod
    public void setup(PluginCall call) {
        String channelId = call.getString("channelId", "");

        if (channelId.length() == 0) {
            call.reject("Must provide a Channel Id");
            return;
        }

        JSObject options = new JSObject();
        options.put("channelId", channelId);
        implementation.setup(options);

        call.resolve();
    }

    @PluginMethod
    public void login(PluginCall call) {
        startActivityForResult(call, implementation.getLoginIntent(), "loginResult");
    }

    @ActivityCallback
    private void loginResult(PluginCall call, ActivityResult result) {
        if (call == null) {
            return;
        }

        JSObject ret = implementation.getLoginResultFromIntent(result.getData());

        returnPluginCall(call, ret);
    }

    @PluginMethod
    public void logout(PluginCall call) {
        JSObject ret = implementation.logout();

        returnPluginCall(call, ret);
    }

    @PluginMethod
    public void getAccessToken(PluginCall call) {
        JSObject ret = implementation.getAccessToken();

        returnPluginCall(call, ret);
    }

    @PluginMethod
    public void refreshAccessToken(PluginCall call) {
        JSObject ret = implementation.refreshAccessToken();

        returnPluginCall(call, ret);
    }

    @PluginMethod
    public void verifyAccessToken(PluginCall call) {
        JSObject ret = implementation.verifyAccessToken();

        returnPluginCall(call, ret);
    }

    private void returnPluginCall(PluginCall call, JSObject ret) {
        if (ret.getString("code") == "SUCCESS") {
            call.resolve(ret);
        } else {
            call.reject(ret.getString("message"), ret.getString("code"));
        }
    }
}

File: ios/Plugin/LineLoginPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(LineLoginPlugin)
public class LineLoginPlugin: CAPPlugin {
    private var implementation: LineLogin?

    override public func load() {
        self.implementation = LineLogin(self)
    }

    @objc func setup(_ call: CAPPluginCall) {
        let channelId = call.getString("channelId") ?? ""

        if channelId.isEmpty {
            call.reject("Must provide a Channel Id")
            return
        }

        implementation?.setup([ "channelId": channelId ])

        call.resolve()
    }

    @objc func login(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.implementation?.login() {result in
                self.returnPluginCall(call, result)
            }
        }
    }

    @objc func logout(_ call: CAPPluginCall) {
        self.implementation?.logout() {result in
            self.returnPluginCall(call, result)
        }
    }

    @objc func getAccessToken(_ call: CAPPluginCall) {
        self.implementation?.getAccessToken() {result in
            self.returnPluginCall(call, result)
        }
    }

    @objc func refreshAccessToken(_ call: CAPPluginCall) {
        self.implementation?.refreshAccessToken() {result in
            self.returnPluginCall(call, result)
        }
    }

    @objc func verifyAccessToken(_ call: CAPPluginCall) {
        self.implementation?.verifyAccessToken() {result in
            self.returnPluginCall(call, result)
        }
    }

    func returnPluginCall(_ call: CAPPluginCall, _ result: [String: Any]) {
        if result["code"] as! String == "SUCCESS" {
            call.resolve(result)
        } else {
            call.reject(result["message"] as! String, result["code"] as? String)
        }
    }
}
